### PR TITLE
Multiple improvements to global router

### PR DESCRIPTION
**Pull request by G10h4ck**
Verify messages hop by hop, this avoid the spreading of invalid messages through
      the network, as a side effect missing RsGxsId keys are propagated hop by hop
      from the sender toward the destination of the message.
    Temporarly store messages which validity could not be verified because of
      missing key in a queue, attempt request sender key and signature validation
      validity periodically until timeout, drop them after timeout, this should
      avoid messages sent by new users being dropped by the GRouter, which is very
      counterintuitive and frustrating for newcomers. The queue is not persistent
      accross RetroShare restarts, so messages not yet validated are lost when
      RetroShare is turned off.
    Do not drop forwarded messages that have unkown service id.
    When receiving a message with unkown service id inform the sender setting
      specific flag in receipt.
    Use type safe flags for GRouter items.
    RsTickEvent::prev_event_ago age is unsigned!
    p3IdService improve some internal implementations and API.
    Use change assert into static_assert for some checks that can be done at compile
      time.
    Solve a bunch of compiler warnings.
    Deprecate some cruft.

### CODE REVIEW

File: libretroshare/src/grouter/grouteritems.h [hunk 1]
```diff
@@ -83,40 +83,52 @@ class RsGRouterNonCopyableObject
...
     GRouterServiceId service_id ;
     RsTlvKeySignature signature ;		// signs mid+destination_key+state
-	uint32_t flags ; 					// packet was delivered, not delivered, bounced, etc
+
+	/// packet was delivered, not delivered, bounced, etc
+	RsGrouterItemFlags flags;
 };
 
-class RsGRouterGenericDataItem: public RsGRouterAbstractMsgItem, public RsGRouterNonCopyableObject
+struct RsGRouterGenericDataItem:
```
Comment (csoler) [inline libretroshare/src/grouter/grouteritems.h]: RsItems generally are classes, not structs. So we should keep the same type of data.

File: libretroshare/src/grouter/groutertypes.h [hunk 2]
```diff
@@ -21,17 +21,22 @@
...
 #include "retroshare/rsgrouter.h"
 
-class RsGRouterGenericDataItem ;
+struct RsGRouterGenericDataItem;
 class RsGRouterSignedReceiptItem ;
 
-static const uint16_t GROUTER_CLIENT_ID_MESSAGES     = 0x1001 ;
+#ifndef V07_NON_BACKWARD_COMPATIBLE_CHANGE_GROUTER_SERVICE_ID
+/// Pay special attention fixing this, as may break retrocompatibility!
+RS_DEPRECATED_FOR(RsServiceType::MSG)
```
Comment (csoler) [inline libretroshare/src/grouter/groutertypes.h:37]: MSG? Probably a copy/paste error?

File: libretroshare/src/grouter/p3grouter.cc [hunk 3]
```diff
@@ -1485,132 +1536,204 @@ bool p3GRouter::sliceDataItem(RsGRouterAbstractMsgItem *item,std::list<RsGRouter
...
+			Dbg2() << __PRETTY_FUNCTION__ << "  message is not at origin. "
+			       << "Setting message transmission to RECEIPT_OK" << std::endl;
 
-    // also note the incoming route in the routing matrix
+			it->second.data_status  = RS_GROUTER_DATA_STATUS_RECEIPT_OK;
+			it->second.receipt_item = receiptItem.release();
+			// no need of dluplicate() just release unique_ptr
+		}
 
-    if(!mTurtle->isTurtlePeer(receipt_item->PeerId()))
```
Comment (csoler) [inline libretroshare/src/grouter/p3grouter.cc:1604]: Why is this removed?

File: libretroshare/src/grouter/groutertypes.h (hunk 2 above)
Comment (G10h4ck) [reply libretroshare/src/grouter/groutertypes.h:37]: this doesn't mean that `RsServiceType::MSG` is deprecated, but that the near defined symbol `GROUTER_CLIENT_ID_MESSAGES` is deprecated and that `RsServiceType::MSG` should be used instead of that

File: libretroshare/src/grouter/p3grouter.cc (hunk 3 above)
Comment (G10h4ck) [reply libretroshare/src/grouter/p3grouter.cc:1604]: It hasn't been removes it, has been moved at line 1649, because this method has been changed a plain diff won't help much

File: libretroshare/src/services/p3gxscircles.cc [hunk 4]
```diff
@@ -1118,7 +1118,7 @@ bool p3GxsCircles::cache_request_load(const RsGxsCircleId &id)
 		return true;
 	}
 
-	int32_t age = 0;
+	uint32_t age = 0;
 	if (RsTickEvent::prev_event_ago(CIRCLE_EVENT_CACHELOAD, age))
```
Comment (csoler) [inline libretroshare/src/services/p3gxscircles.cc:1122]: Technically it's not possible to ensure that age will not be negative. That is why this function uses a int32 instead of a uint32_t. So hanging to int32 is probably bad.

File: libretroshare/src/util/rstickevent.h [hunk 5]
```diff
@@ -48,8 +49,8 @@ void    schedule_event(uint32_t event_type, rstime_t when, const std::string &el
 void    schedule_in(uint32_t event_type, uint32_t in_secs);
 void    schedule_in(uint32_t event_type, uint32_t in_secs, const std::string &elabel);
 
-int32_t event_count(uint32_t event_type);
-bool 	prev_event_ago(uint32_t event_type, int32_t &age);
+    int32_t event_count(uint32_t event_type);
+	bool prev_event_ago(uint32_t event_type, uint32_t& age);
```
Comment (csoler) [inline libretroshare/src/util/rstickevent.h:53]: probably not a good idea. See above comment.

File: retroshare.pri [hunk 6]
```diff
@@ -262,6 +271,7 @@ rs_v07_changes {
     DEFINES += V07_NON_BACKWARD_COMPATIBLE_CHANGE_003
     DEFINES += V07_NON_BACKWARD_COMPATIBLE_CHANGE_004
     DEFINES += V07_NON_BACKWARD_COMPATIBLE_CHANGE_UNNAMED
+    DEFINES += V07_NON_BACKWARD_COMPATIBLE_CHANGE_GROUTER_SERVICE_ID
```
Comment (csoler) [inline retroshare.pri:274]: What is the need for that particular change?

File: libretroshare/src/retroshare/rsgrouter.h [hunk 7]
```diff
@@ -19,19 +20,29 @@
...
+
+RS_DEPRECATED_FOR(RsGxsId)
+typedef RsGxsId GRouterKeyId;
+
+/**
+ * @note Some items use this type, when migrating to RsServiceType (2 bytes),
+ * special care should be taken to do it in a way which doesn't break
+ * retrocompatibility
+ */
+RS_DEPRECATED_FOR(RsServiceType)
```
Comment (csoler) [inline libretroshare/src/retroshare/rsgrouter.h:39]: I dont understand that. The GRouterServiceId is rather a client ID. Such a client is not necessarily a RSService. So merging the two limit possibilities and does not help much either.

File: libretroshare/src/rsitems/rsserviceids.h [hunk 8]
```diff
@@ -40,7 +40,6 @@ enum class RsServiceType : uint16_t
 	FILE_DATABASE              = 0x0019,
 	SERVICEINFO                = 0x0020,
 	BANDWIDTH_CONTROL          = 0x0021,
-	MAIL                       = 0x0022,
```
Comment (csoler) [inline libretroshare/src/rsitems/rsserviceids.h:43]: If possible keep the line with "unused since XXX (formerly MAIL)". So when we use this number, we know that we dont have risks to collide with an existing service id that was only recently removed.

File: libretroshare/src/services/p3gxscircles.cc (hunk 4 above)
Comment (csoler) [reply libretroshare/src/services/p3gxscircles.cc:1122]: I'm telling this because, even if the logic ensures that it behaves in some way, a bug can turn this number into a < 0 number and therefore make it way too large. If you can avoid that when it is computed, then it's fine.

File: libretroshare/src/services/p3msgservice.cc [hunk 9]
```diff
@@ -1988,13 +1997,38 @@ void p3MsgService::manageDistantPeers()
...
-                mGRouter->registerKey(*it,GROUTER_CLIENT_ID_MESSAGES,"Messaging contact") ;
-            else
-                mGRouter->unregisterKey(*it,GROUTER_CLIENT_ID_MESSAGES) ;
-        }
+		for( auto it = std::as_const(own_id_list).begin();
+		     it != own_id_list.end(); ++it )
+		{
+			if(mShouldEnableDistantMessaging)
+			{
+#ifndef V07_NON_BACKWARD_COMPATIBLE_CHANGE_GROUTER_SERVICE_ID
```
Comment (csoler) [inline libretroshare/src/services/p3msgservice.cc:2005]: clearly at this point we see that this change is problematic because it adds unnecessary ifdefs and retrocompatibility issues.

File: libretroshare/src/services/p3msgservice.cc [hunk 10]
```diff
@@ -1988,13 +1997,38 @@ void p3MsgService::manageDistantPeers()
...
+				// Keep for RetroShare < 0.6.6 retro-compatibility
+				mGRouter->registerKey(
```
Comment (csoler) [inline libretroshare/src/services/p3msgservice.cc:2007]: Same here! You get to register 2 GRouter clients in order to keep backward compat!

File: libretroshare/src/grouter/groutertypes.h (hunk 2 above)
Comment (csoler) [reply libretroshare/src/grouter/groutertypes.h:37]: That's an error. See my other comments about GRouterId type.

File: libretroshare/src/grouter/p3grouter.cc [hunk 11]
```diff
@@ -3,7 +3,8 @@
  *                                                                             *
  * libretroshare: retroshare core library                                      *
  *                                                                             *
- * Copyright 2013 by Cyril Soler <csoler@users.sourceforge.net>                *
+ * Copyright (C) 2013  Cyril Soler <csoler@users.sourceforge.net>              *
+ * Copyright (C) 2019  Gioacchino Mazzurco <gio@eigenlab.org>                  *
```
Comment (csoler) [inline libretroshare/src/grouter/p3grouter.cc:7]: Personnally I dont add my own name everytime I modify a file, unless for a full rewrite, which means the file can be considered as new. For small changes, individual commits hold your name already and git blame give the exact spectrum of your contributions.

**Comment by csoler** [comment]
also I would really prefer not to see simple memory management turned into a forest of std::unique_ptr. These things are nice are useful, except when not using them is much simpler. A typical example is in the handleIncoming method(). All methods called below take const RsItems* and do not delete them. They are all deleted in the parent. Turning all this into std::unique_ptr makes it totally impossible to follow/read.

**Comment by csoler** [comment]
... and it makes the important changes (mostly the queuing of unhandled items) totally impossible to spot and check in the PR.

**Comment by G10h4ck** [comment]
Superseded by other smaller pull requests
